### DISCUSSION TITLE
Problem with the Google sign in

### DISCUSSION THREAD

**Discussion by baudee** [serverpod/serverpod#1019] [Q&A] [answered]
I implemented the Google sign in with serverpod. When I click on the google button, I can choose my google account. After that, a loading widget appears and after few seconds I get a black screen with the error below. The sign in works because I have a new row in my db and when I reload the app, I am authenticated. The problem comes from a pop, without screen in the stack. Does anyone knows how to fix that ?

I/flutter (29254): serverpod_auth_google: Signing out from google
I/flutter (29254): serverpod_auth_google: PlatformException(status, Failed to disconnect., null, null)
I/flutter (29254): serverpod_auth_google: Authentication was successful. Saved credentials to SessionManager
E/flutter (29254): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 'package:go_router/src/delegate.dart': Failed assertion: line 145 pos 7: '_matchList.isNotEmpty': You have popped the last page off of the stack, there are no pages left to show
E/flutter (29254): #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
E/flutter (29254): #1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
E/flutter (29254): #2      GoRouterDelegate._debugAssertMatchListNotEmpty (package:go_router/src/delegate.dart:145:7)
E/flutter (29254): #3      GoRouterDelegate._handlePopPageWithRouteMatch.<anonymous closure> (package:go_router/src/delegate.dart:163:7)
E/flutter (29254): #4      GoRouterDelegate._handlePopPageWithRouteMatch (package:go_router/src/delegate.dart:165:6)
E/flutter (29254): #5      _PagePopContext.onPopPage (package:go_router/src/builder.dart:622:35)
E/flutter (29254): #6      NavigatorState.pop (package:flutter/src/widgets/navigator.dart:4993:28)
E/flutter (29254): #7      SignInWithGoogleButtonState.build.<anonymous closure>.<anonymous closure> (package:serverpod_auth_google_flutter/src/signin_button.dart:84:33)

**Comment by Isakdl** [comment]
Could you give a minimal code example for the issue? I can see that you are using `go router`, this issue could perhaps have something to do with how go router interacts with the native routing functionality.

The Google Sign in button normally pushes a loading screen to the stack and then pops it after a successful/failed login attempt.

**Reply by baudee** [reply]
So I'm using this button :  
 SignInWithGoogleButton(
    caller: client.modules.auth,
    redirectUri: Uri.parse('http://localhost:8082/googlesignin'),
    serverClientId: "...",
    onSignedIn: () {
      print('Signed in');
    },
    onFailure: () {
      print('Failed to sign in');
    }
  )
  
  And here is my router :
  class AppRouter {
  final listenable = ValueNotifier(sessionManager.isSignedIn);

  GoRouter get router => _goRouter;

  // Singleton with parameter this.appService
  AppRouter._internal(){
    sessionManager.addListener(() async {
      await Future.delayed(const Duration(seconds: 5));
      listenable.value = sessionManager.isSignedIn;
    });
  }

  factory AppRouter() {
    _instance ??= AppRouter._internal();
    return _instance!;
  }
  static AppRouter? _instance;

  late final GoRouter _goRouter = GoRouter(
    refreshListenable: listenable,
    initialLocation: HomePage.routeName,
    routes: [
      ShellRoute(
        builder: (context, state, child) {
          return AppView(child: child);
        },
        routes: [...],
      ),
    ],
  );
}

**Comment by rodjordan** [comment]
If you are using a redirect listenable in GoRouter refreshListenable: to watch your logged in status to redirect your application  after a login it may be picking up the login state change at the same time you are trying to pop the login screen. If you put a delay of 1 or 2 seconds into your redirection listenable like the sample below, see if that proved the point...At least that may help you find the issue....

class WrapFirebaseAuth extends ChangeNotifier {
  final fbauth.FirebaseAuth fba;
  late StreamSubscription<fbauth.User?> str;

  WrapFirebaseAuth(this.fba) {
    notifyListeners();
    str= fba.authStateChanges().asBroadcastStream().listen((event) async {
      await Future.delayed(const Duration(seconds: 3));
      notifyListeners();
    });
  }

**Reply by baudee** [reply]
I'm using a refreshListenable. Tried with a 5 seconds delay, but still get the same error

**Reply by rodjordan** [reply]
I can reproduce the same error as you but not every time. (which makes me think it is a redirection race)
However, I noticed (by adding logging) that the GoRouter redirect (triggered by the sessionManager listenable) is being called before the onSignIn callback of the SignInWithGoogleButton. I am not sure if this helps, but I think it could be a side effect of the redirect listenable not playing nice with the post sign in redirect of the SignInWithGoogleButton. Have you tried removing the redirect from the SignInWithGoogleButton to let the listenable do the work of redirecting?

**Reply by baudee** [reply]
redirectUri is required in the SignInWithGoogleButton, so I cannot remove it. I don't really know what else to try. Is there an alternative to refreshListenable for go_router ?

**Reply by baudee** [reply, marked as answer]
Resolved it by using signInWithGoogle method in a custom button instead of SignInWithGoogleButton widget
